platform-sdk/docs/proposals/001-merkle-state-api/001-merkle-state-api.md

```diff
@@ -2,34 +2,72 @@
 
 ## Purpose and Context
 
+### Why?
+
+The Merkle tree is a rather complex data structure, and to make interaction easier, we need to hide this complexity behind the API.
+Another reason to have this API is to provide additional flexibility in changing the implementation details of the Merkle tree.
+As long as they are hidden, we are free to change them.
+
+Note that this API is already implemented as a part of the Service Modularization project, but these classes will be moved
+to a separate module as a part of this proposal.
+
+### Context
+
 There are two issues that need to be addressed:
 
 * We need to have a public API for the Merkle tree that is simple and easy to use, residing in a separate module with a minimal set of dependencies.
 * The Merkle tree has to be fully expressed in terms of State API. Currently, `PlatformState` stands out and needs to be refactored.
 
-The goal of this design proposal is to ensure that one can interact with all the objects in the Merkle tree using the newly introduced State API.
+#### Singleton
+
+A singleton is a node with 2 children:
 
-This refactoring will enable optimized state proof, as it will be possible to represent an n-ary root node as a virtual binary tree.
-Therefore, the number of hashes in the state proof will be minimized.
+- string label for State Name
+- protobuf message for child data
+
+In Java code it's represented by the following classes:
+```java
+public interface ReadableSingletonState<T> {
+  @NonNull
+  String getStateKey();
+  T get();
+  boolean isRead();
+}
+
+public interface WritableSingletonState<T> extends ReadableSingletonState<T> {
+  void put(@Nullable T value);
+
+  boolean isModified();
+}
+```
 
 ## Requirements
 
 - `PlatformState` should be refactored to a singleton object as defined by the State API.
-- the Merkle tree should be reshaped so that `MerkleHederaState` becomes the root of the tree and `PlatformState` becomes its child.
 - there should be a new module - `swirlds-state-api` - to host a set of interfaces, records, and abstract classes that represent the Hedera state.
-  This module should have a minimal set of dependencies. Ideally, the Block Node should not have a compile-time dependency
+  This module should have a minimal set of dependencies. The Block Node should not have a compile-time dependency
   on any other modules but this one to interact with the state.
-- there should be a new module - `swirlds-state-impl` module to host the implementation of these interfaces and helper classes.
 
 ## Changes
 
 ### Public API
 
-State API classes and interfaces will migrate to a designated module. The classes themselves will not change.
+State API classes and interfaces will migrate to a designated module. The classes themselves will not change. `PlatformState` class will become a part of the public API.
+
+Usage example:
+```java
+private PlatformState findPlatformState(HederanState state) {
```

Comment:
Nit: s/Hederan/Hedera/

Reply:
Better use hashgraph

Reply:
@hendrikebbers would you recommend renaming interface `com.swirlds.state.HederaState` into `com.swirlds.state.HashgraphState`?

Reply:
If it is a new one, yes

Reply:
It's not a new one. We have a `HederaState` interface now.